File: 54_Letter/javascript/letter.js
// LETTER
//
// Converted from BASIC to Javascript by Oscar Toledo G. (nanochess)
//

function print(str)
{
    document.getElementById("output").appendChild(document.createTextNode(str));
}

function input()
{
    var input_element;
    var input_str;

    return new Promise(function (resolve) {
                       input_element = document.createElement("INPUT");

                       print("? ");
                       input_element.setAttribute("type", "text");
                       input_element.setAttribute("length", "50");
                       document.getElementById("output").appendChild(input_element);
                       input_element.focus();
                       input_str = undefined;
                       input_element.addEventListener("keydown", function (event) {
                                                      if (event.keyCode == 13) {
                                                      input_str = input_element.value;
                                                      document.getElementById("output").removeChild(input_element);
                                                      print(input_str);
                                                      print("\n");
                                                      resolve(input_str);
                                                      }
                                                      });
                       });
}

function tab(space)
{
    var str = "";
    while (space-- > 0)
        str += " ";
    return str;
}

// Main program
async function main()
{
    print(tab(33) + "LETTER\n");
    print(tab(15) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n");
    print("\n");
    print("\n");
    print("\n");
    print("LETTER GUESSING GAME\n");
    print("\n");
    print("I'LL THINK OF A LETTER OF THE ALPHABET, A TO Z.\n");
    print("TRY TO GUESS MY LETTER AND I'LL GIVE YOU CLUES\n");
    print("AS TO HOW CLOSE YOU'RE GETTING TO MY LETTER.\n");
    while (1) {
        l = 65 + Math.floor(26 * Math.random());
        g = 0;
        print("\n");
        print("O.K., I HAVE A LETTER.  START GUESSING.\n");
        while (1) {

            print("\n");
            print("WHAT IS YOUR GUESS");
            g++;
            str = await input();
            a = str.charCodeAt(0);
            print("\n");
            if (a == l)
                break;
            if (a < l) {
                print("TOO LOW.  TRY A HIGHER LETTER.\n");
            } else {
                print("TOO HIGH.  TRY A LOWER LETTER.\n");
            }
        }
        print("\n");
        print("YOU GOT IT IN " + g + " GUESSES!!\n");
        if (g > 5) {
            print("BUT IT SHOULDN'T TAKE MORE THAN 5 GUESSES!\n");
        } else {
            print("GOOD JOB !!!!!\n");
        }
        print("\n");
        print("LET'S PLAY AGAIN.....");
    }
}

main();

File: 00_Alternate_Languages/54_Letter/ruby/letter.rb
#!/usr/bin/env ruby

# Kinema
# reinterpreted from BASIC by stephan.com

puts 'Letter'.center(80)
puts 'Adapted by stephan.com'.center(80)
puts "\n\n\n"

puts "Letter guessing game\n\n"

puts "I'll think of a letter of the alphabet, A to Z."
puts "Try to guess my letter and I'll give you clues"
puts "as to how close you're getting to my letter."

def win(turns)
  puts "\nyou got it in #{turns} guesses!!"
  return puts "but it shouldn't take more than 5 guesses!" if turns > 5

  puts "good job !!!!!\a\a\a"
end

def play
  letter = ('A'..'Z').to_a.sample
  guess = nil
  turn = 0

  puts "\nO.K., I have a letter.  Start guessing."

  until guess == letter
    puts "\nWhat is your guess?"

    guess = gets.strip.chars.first.upcase
    turn += 1

    puts 'Too low.  Try a higher letter.' if guess < letter
    puts 'Too high.  Try a lower letter.' if guess > letter
  end
  win(turn)
end

loop do
  play
  puts "\nlet's play again....."
end
